runtime/compiler/trj9/runtime/J9Profiler.cpp

```diff
@@ -894,29 +881,50 @@ TR_ValueProfileInfoManager::TR_ValueProfileInfoManager(TR::Compilation *comp)
    }
 
 TR_AbstractInfo *
-TR_ValueProfileInfoManager::getValueInfo(TR::Node *node, TR::Compilation *comp, uint32_t profileInfoKind, TR_ValueInfoType type)
+TR_ValueProfileInfoManager::getProfiledValueInfo(TR::Node *node, TR::Compilation *comp, TR_ValueInfoKind kind,
+      uint32_t source)
    {
-   return getValueInfo(node->getByteCodeInfo(), comp, profileInfoKind, type);
+   TR_ValueProfileInfoManager *manager = TR_ValueProfileInfoManager::get(comp);
+   return manager ? manager->getValueInfo(node, comp, kind, source) : NULL;
    }
 
 TR_AbstractInfo *
-TR_ValueProfileInfoManager::getValueInfo(TR_ByteCodeInfo &bcInfo, TR::Compilation *comp, uint32_t profileInfoKind, TR_ValueInfoType type)
+TR_ValueProfileInfoManager::getProfiledValueInfo(TR_ByteCodeInfo &bci, TR::Compilation *comp, TR_ValueInfoKind kind,
+      uint32_t source)
+   {
+   TR_ValueProfileInfoManager *manager = TR_ValueProfileInfoManager::get(comp);
+   return manager ? manager->getValueInfo(bci, comp, kind, source) : NULL;
+   }
+
+TR_AbstractInfo *
+TR_ValueProfileInfoManager::getValueInfo(TR::Node *node, TR::Compilation *comp, TR_ValueInfoKind kind,
+      uint32_t source)
+   {
+   return getValueInfo(node->getByteCodeInfo(), comp, kind, source);
+   }
+
+TR_AbstractInfo *
+TR_ValueProfileInfoManager::getValueInfo(TR_ByteCodeInfo &bcInfo, TR::Compilation *comp, TR_ValueInfoKind kind,
+      uint32_t source)
    {
    TR_AbstractInfo *info = NULL;
+   bool internal = _jitValueProfileInfo && (source == allProfileInfo || source == justJITProfileInfo);
+   bool external = source == allProfileInfo || source == justInterpreterProfileInfo;
 
-   if (_jitValueProfileInfo && (profileInfoKind != justInterpreterProfileInfo))
-      info = _jitValueProfileInfo->getValueInfo(bcInfo, comp, type);
+   if ((!info || info->getTotalFrequency() == 0) && internal)
```

Comment:
in general it would be more neat to do internal && (....). That would have the benefit of testing the already computed bool first before potentially doing a deref which could be a savings if this runs a lot.